### PR TITLE
The case of the missing ada

### PR DESCRIPTION
**Pull request by nc6**
A fun one, this.

The scenario:

In one of our tests, 163,000 ADA vanished, violating the preservation of ADA property. How did this happen? The sequence of events was something like this:

1. Key 24583f7b is registered, and is set as the reward account for a pool.
2. The pool produces a number of blocks.
3. A reward update is created which includes a payment of 163765411781 lovelace to 24583f7b.
4. Key 24583f7b is deregistered.
5. The reward update created in 3 is applied. This has the effect of adding 24583f7b back into the `_rewards` map, but not the `_stkCreds` map.
6. Key 24583f7b is re-registered. Since it is not in the `_stkCreds` map, this is allowed. However, the registration zeros its rewards, resulting in the destruction of those 163,000 ADA.

Why did we not spot this sooner? Well, because previously we had:
```
_rewards = _rewards ds ∪ Map.singleton (RewardAcnt network hk) (Coin 0)
``` 

during certificate registration. But during the performance upgrades (in d716ad) this was changes to

```_rewards = addpair (RewardAcnt network hk) (Coin 0) (_rewards ds),```

Union is left-biased, and hence will not overwrite. addpair is implemented to ignore the old value and overwrite it with the new. Hence before d716ad, the reward account still contained funds, albeit inaccessible ones.

This PR does a number of things:

- It expands the test for ADA preservation to be more helpful.
- It adds an additional check to the ADA preservation that verifies a more local preservation property - that outside the epoch boundary, the change in rewards must be equal to the sum of the withdrawls.
- It adds another test which verifies that the `stkCreds` and `rewards` maps stay in sync (e.g. that they have the same domain).
- It adds a sanity check during key registration that the key is not already in `rewards`.
- It fixes `applyRUpd` by checking whether the keys are still registered. Any funds allocated to deregistered keys are now returned to the reserves.

### CODE REVIEW

**Review by uroboros — APPROVED** [review]
BRAVO!

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs [hunk 1]
```diff
@@ -222,6 +232,7 @@ delegationTransition = do
       -- note that pattern match is used instead of regCred, as in the spec
       -- hk ∉ dom (_stkCreds ds) -- Specification code translates below
       not (haskey hk (_stkCreds ds)) ?! StakeKeyAlreadyRegisteredDELEG hk
+      not (haskey (RewardAcnt network hk) (_rewards ds)) ?! StakeKeyInRewardsDELEG hk
```
Comment (JaredCorduan) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs:235]: want to add this change to the spec too? I can do it if that would be helpful.

**Review by JaredCorduan — APPROVED** [review]
:clap: :clap: :detective:

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs (hunk 1 above)
Comment (nc6) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Deleg.hs:235]: I think this change isn't too important (it's a sanity check), but if you felt like adding the `applyRUpd` changes to the spec that would be great!

**Pull request merged**